### ISSUE TITLE
Window dragging sometimes fails

### ISSUE THREAD

**Issue by CCatington** [clsid2/mpc-hc#1314]
(Copied more or less verbatim from https://codecs.forumotion.net/t3888-window-dragging-sometimes-fails)

I recently updated MPC-HC from 1.9.13.14 to 1.9.15.76 (KLCP 16.2.0 to 16.4.0) and got the issue that dragging MPC-HC's window by clicking inside it (so not by using the titlebar) sometimes does nothing, no matter how far I drag.
It seems very inconsistent but when I find a point where it fails, trying to click-drag close to it will almost always fail too, almost as if there's some invisible rectangle that's being dragged around or something weird like that.
This happens with MPC-HC 1.9.15.79 (KLCP 16.4.6) too and for me this is very noticable since I always shuffle the window around like this, though trying to reproduce it on the spot can take quite a few tries.

Looking at the changelog I found

> Added a small safezone above and below the seekbar where (accidental) mouse clicks/drags are ignored.

That seems a likely candidate, here's a link to the PR: #1171 (and also #1177).
I can't really tell if there's something obviously wrong there, but it's the only lead I got so far.

**Comment by adipose** [comment]
Are you clicking near the seekbar when you try to drag?

**Comment by CCatington** [comment]
No, it can happen everywhere.
Quoting the details from the forum thread:

> I can reproduce it dragging anywhere like right in the middle of the window and without the window being maximized.
> I think I even had it happen that I couldn't click "Play" or "Navigate" once, but I can't say for sure since that happened so fast.
>
> I recorded it in action here:
>
> ![Ntbotfu](https://user-images.githubusercontent.com/91107487/135845170-24099618-170f-4c58-a96e-2f2881a9c233.gif)
>
> I first found a spot by trying to drag the window at random positions (which can take 50+ tries) and once found, click-dragging that spot consistently fails.
Once I drag it somewhere else, the deadzone seems to move to another random spot.
The deadzone seems to be more wide than tall, so I still think it's related to the seekbar deadzone.
> 
> What's really odd to me however is this: I can confirm that the seekbar deadzone is definitely working (even downgraded to KLCP 16.2.0 to double check), so it seems the issue is not simply that the seekbar deadzone is placed wrong, so maybe the bug is actually caused by something else, maybe somehow related to window messaging?

At first I thought it was some platform dependent bug but the KLCP admin tried on Win7 just like me and couldn't reproduce the issue, so I'm really not sure what's going on.

Some more technical details:

> tried changing a few settings like DefaultToolbarSize, the video output, also tried in compact mode but alas...
> can reproduce on 1.9.13.37 (took a while) but not on 1.9.13.21 as expected (tried like 200 times).

**Comment by clsid2** [comment]
I think I found the cause.

isSafeZone() is called twice when clicking on the video area. First from CMouse::InternalOnLButtonDown with point relative to video window, then in CMainFrame::WindowProc with point relative to screen.

**Comment by clsid2** [comment]
In CMainFrame::WindowProc the point is relative to the screen when clicking on the video area, and relative to player window when clicking elsewhere.

**Comment by clsid2** [comment]
Found a fix.

**Comment by CCatington** [comment]
Got some good and some bad news.

The bad news is that I tested dev build 1.9.16.19 (which according to the version string contains 9c7f217) and I can still reproduce the issue, so looks like this wasn't it.

The good news is that - after quite some trouble - I was finally able to compile mpc-hc so I can hopefully help out by debugging the issue firsthand.
Surely not today though :zzz:

**Comment by CCatington** [comment]
Okay, finally found some time to look into it!

So first things first:
According to [the docs](https://docs.microsoft.com/en-us/windows/win32/inputdev/wm-nclbuttondown), when `CMainFrame::WindowProc()` is called with a `WM_NCLBUTTONDOWN` message, `lParam` is supposed to contain screen coordinates:

> lParam
> A `POINTS` structure that contains the x- and y-coordinates of the cursor. The coordinates are relative to the upper-left corner **of the screen**.

This means there should be a `ScreenToClient(&pt);` before the `isSafeZone()` call there, since the latter works with relative coords.
The `CMouse::InternalOnLButtonDown()` side should be perfectly fine as-is.

This however also means that what you mentioned earlier:

> In CMainFrame::WindowProc the point is relative to the screen when clicking on the video area, and relative to player window when clicking elsewhere.

should not be happening!
I found that (after undoing 9c7f217e8e274eb4beba47879722a648fbc9dda4) `CMainFrame::WindowProc()` gets absolute coords when clicking the title bar, the navbar, the video area, ... but not when clicking any of the bottom elements like the seek bar, the controls or status bar.
This defies the definitions, so you know something is up.

Anyone got an idea what might be causing this?

---

On a sidenote:
The docs also mention to not use the `LOWORD` or `HIWORD` macros (which are used inside `POINTSTOPOINT`) because they can't deal with negative coordinates on multi-monitor systems but instead `GET_X_LPARAM`, `GET_Y_LPARAM` should be used.

**Comment by clsid2** [comment]
@adipose
Any idea why the point value isn't relative to screen when clicking below the seekbar? Maybe some hidden bug in the theme code? The call stack does not show any clues. Maybe of of the uses of ScreenToClient() in the code does something with permanent effect, but I don't see anything obvious that is wrong on a quick look.

With GetCursorPos() we can get get value that is near the original click. That could be used to check if the point value is relative to the screen or not. But that is a nasty workaround. I rather discover the actual cause of the problem.

**Comment by adipose** [comment]
I will take a look. Probably something funny with the mouse window code.

**Comment by clsid2** [comment]
Workaround code:
```
    if (message == WM_NCLBUTTONDOWN && wParam == 2) {
        CPoint pt = CPoint(GET_X_LPARAM(lParam), GET_Y_LPARAM(lParam));
        CPoint pt2 = pt;
        GetCursorPos(&pt2);
        TRACE(_T("WindowProc: x=%d y=%d Mouse: x=%d y=%d\n"), pt.x, pt.y, pt2.x, pt2.y);
        if (abs(pt.x-pt2.x) < 25 && abs(pt.y - pt2.y) < 25) {
            ScreenToClient(&pt);
            TRACE(_T("ScreenToClient: x=%d y=%d\n"), pt.x, pt.y);
        }
        if (isSafeZone(pt)) {
            return 0;
        }
    }
```

**Comment by CCatington** [comment]
Ah, I think I found it!
There actually seem to be a lot of places where relative coordinates are passed to `WindowProc`.

Here's for the seek bar:
https://github.com/clsid2/mpc-hc/blob/29a1254b5d4cadd3916b3ae4d3456bc3410af213/src/mpc-hc/PlayerSeekBar.cpp#L740-L741
Replacing `MapWindowPoints(...)` with `ClientToScreen(&point);` fixes it.

Same goes for the toolbar, info bar and status bar:
https://github.com/clsid2/mpc-hc/blob/29a1254b5d4cadd3916b3ae4d3456bc3410af213/src/mpc-hc/PlayerToolBar.cpp#L476-L477
https://github.com/clsid2/mpc-hc/blob/29a1254b5d4cadd3916b3ae4d3456bc3410af213/src/mpc-hc/PlayerInfoBar.cpp#L286-L287
https://github.com/clsid2/mpc-hc/blob/29a1254b5d4cadd3916b3ae4d3456bc3410af213/src/mpc-hc/PlayerStatusBar.cpp#L466-L471

There might be some I missed.
This is also only for the `WM_NCLBUTTONDOWN` event, so some other events might have the same issue - I don't think it matters now, but it might in the future.

(It also looks like `MAKELPARAM` can deal with negative coords, so everything should work multi-monitor too.)

**Comment by clsid2** [comment]
Good find!
We should probably check all uses of PostMessage() that involve coordinates. I will wait until adipose can confirm if this fix doesn't break anything. He is more familiar with the GUI code.

**Comment by CCatington** [comment]
**tl;dr** Couldn't find any more issues.

Yeah, need to be pretty careful with this stuff, there's e.g. the `WM_LBUTTONDOWN` event (notice the missing `NC` "not captured") which actually takes [relative coords](https://docs.microsoft.com/en-us/windows/win32/inputdev/wm-lbuttondown#parameters) just to spite you. At first I thought there were some potential bugs with some [madVR compatibility code](https://github.com/clsid2/mpc-hc/blob/29a1254b5d4cadd3916b3ae4d3456bc3410af213/src/mpc-hc/MouseTouch.cpp#L231-L266) - until I looked it up.

I did another sweep to check if I missed something.
Searching for `WindowProc`, there are some calls to `CallWindowProc()` and `__super::WindowProc()` but nothing seems to be dealing with coordinates.
There's this part inside `CMainFrame::WindowProc` that will be affected by the suggested change:
https://github.com/clsid2/mpc-hc/blob/3123f73836e3b77be432d645157a97b95cf82620/src/mpc-hc/MainFrm.cpp#L18790-L18794
The first call passes messages to the madVR window (for hotkey handling I assume), but I don't think madVR even handles mouse events and the latter just calls the default `CWnd::WindowProc`, so neither should cause a new issue.

Searching for `WM_NC` turns up no other event except `WM_NCLBUTTONDOWN` and there is actually one more of them here:
https://github.com/clsid2/mpc-hc/blob/29a1254b5d4cadd3916b3ae4d3456bc3410af213/src/mpc-hc/MouseTouch.cpp#L496
but I'm pretty sure `m_beginDragPoint` is already absolute since it's being compared to a point called `screenPoint`.

Searching for `PostMessage` doesn't seem to turn anything new up either.

So looks like the 4 cases above are already all that need to be fixed and it seems to be safe to do so, though some more eyes on this surely can't hurt.

**Comment by clsid2** [comment]
I have committed the new patch.

**Comment by adipose** [comment]
This looks like the right solution.  MapWindowPoints maps from one relative set of coordinates to another, in another window.  It's pretty useful when the destination requires relative coordinates.  WM_NCLBUTTONDOWN requires screen coordinates, so it's the wrong solution.

I don't remember exactly, but not long ago I reviewed a use of MapWindowPoints and nearly changed it, until I realized it was right.  I'm sure it's related to mousetouch in some way, but perhaps not this exact event.  If I could remember what it was, that would set my mind at ease...

**Comment by CCatington** [comment]
> I have committed the new patch.

Great!
Just one last thing, I was wondering what the ` && wParam == 2` from https://github.com/clsid2/mpc-hc/commit/9c7f217e8e274eb4beba47879722a648fbc9dda4 actually did.

Reading some more docs, `wParam` for `WM_NCLBUTTONDOWN` should [tell you what part of the window was clicked](https://docs.microsoft.com/en-us/windows/win32/inputdev/wm-nclbuttondown#parameters), here's the [list of all possible values](https://docs.microsoft.com/en-us/windows/win32/inputdev/wm-nchittest#return-value).
Value 2 is `HTCAPTION` which means "In a title bar", which is odd since it's also 2 when clicking/dragging the mouse over the video area, seekbar etc.
Looking at the `PostMessage()` calls again, they all set `wParam` to `HTCAPTION` so that's why.

So the conditions effectively filters for title bar or the 4 components.

It'd probably make more sense if those `PostMessage()` calls set `wParam` to `HTCLIENT` (= 1) for "In a client area" and the condition could be changed accordingly - though I'm pretty sure it doesn't actually matter. If the deadzone is positioned right, `wParam` should currently always be 2 if `isSafeZone` is true.

> I don't remember exactly, but not long ago I reviewed a use of MapWindowPoints and nearly changed it, until I realized it was right. I'm sure it's related to mousetouch in some way, but perhaps not this exact event. If I could remember what it was, that would set my mind at ease...

There's one right in `isSafeZone` which is called from mousetouch since https://github.com/clsid2/mpc-hc/commit/994e5872fa15dda9f5f89da0b460a638ab11f9a6, maybe that's the one you're thinking about?

Otherwise this sounds like you need to frantically double check all uses of `MapWindowPoints` again, good luck!
Jokes aside, there's not actually that many `MapWindowPoints` calls now, it's 15 down from 19, so if it's still there it shouldn't be hard to find and to leave a comment á la "this looks wrong but it's actually right because ...".

**Comment by adipose** [comment]
![image](https://user-images.githubusercontent.com/3324395/136457094-1c175fec-37f9-4ed4-8f7b-76798f968c68.png)

I see the code here is 8 years old.  My guess is, it worked fine, because it was trying to measure differential drag.  Now that we're checking the actual point, it's a problem.

Fix looks right.

**Comment by adipose** [comment]
The reason for wParam=2 is title bar is the only place you're normally allowed to drag from.  So we're treating the event as if it were a titlebar drag.

**Comment by CCatington** [comment]
> The reason for wParam=2 is title bar is the only place you're normally allowed to drag from. So we're treating the event as if it were a titlebar drag.

Ohh, so the dragging actually happens through the default handler, am I getting this right?
Either way, put like this, it makes a lot of sense!

**Comment by adipose** [comment]
Also made this tiny patch: https://github.com/clsid2/mpc-hc/pull/1341

**Comment by clsid2** [comment]
The reason I check for `wParam==2` is to prevent doing the safezone check when resizing the window. This could happens if all controls are hidden except for seekbar. When touching the border wParam has a different value (15 if I remember correctly).

**Comment by adipose** [comment]
> 
> 
> > The reason for wParam=2 is title bar is the only place you're normally allowed to drag from. So we're treating the event as if it were a titlebar drag.
> 
> Ohh, so the dragging actually happens through the default handler, am I getting this right? Either way, put like this, it makes a lot of sense!

Yes, if the safezone check succeeds, and madvr doesn't cancel the click, it eventually goes to ::WindowProc(), which should treat it like a titlebar drag.

The other widgets are meant to pass the clicks along to the mainform, which is the code you corrected.  Instead of trapping the events there, they can be used to drag--effectively allowing clicking almost anywhere in the interface and dragging the window.

**Comment by adipose** [comment]
> > I don't remember exactly, but not long ago I reviewed a use of MapWindowPoints and nearly changed it, until I realized it was right. I'm sure it's related to mousetouch in some way, but perhaps not this exact event. If I could remember what it was, that would set my mind at ease...
> 
> There's one right in `isSafeZone` which is called from mousetouch since [994e587](https://github.com/clsid2/mpc-hc/commit/994e5872fa15dda9f5f89da0b460a638ab11f9a6), maybe that's the one you're thinking about?

Yeah, I did look at that one, and I think it's actually the one.  It's definitely right, there.
